poly/schnorr.go

```diff
@@ -0,0 +1,270 @@
+}
+
+// RevealPartialSig reveals the partial signature for this peer
+// Si = Ri + H(m || V) * Pi
+// with :
+// 	- Ri = share of the random secret for peer i
+//  - V  = public commitment of the random secret (i.e. Public random poly evaluated at point 0 )
+//  - Pi = share of the longterm secret for peer i
+// This signature is to be sent to each others peers
+func (s *Schnorr) RevealPartialSig() *PartialSchnorrSig {
+	hash := *s.hash
+	sigma := SUITE.Secret().Zero()
+	sigma = sigma.Add(sigma, *s.random.Share)
+	// H(m||v) * Pi
+	hash = SUITE.Secret().Mul(hash, *s.Longterm.Share)
+	// Ri + H(m||V) * Pi
+	sigma = sigma.Add(sigma, hash)
+
+	psc := &PartialSchnorrSig{
+		Index: s.index(),
+		Part:  &sigma,
+	}
+	return psc
+}
+
+// AddPartialSig receives a signature from others peer,
+// adds it to its list of partial signatures and verify it
+// It return an error if
+// - 	it can not validate this given partial signature
+// 		against the longterm and random shared secret
+// - there is already an partial signature added for this index
+// NOTE : let s = RevealPartialSig(), s is NOT added automatically to the
+// set of partial signature, for now you have to do it yourself by calling
+// AddPartialSig(s)
+func (s *Schnorr) AddPartialSig(ps *PartialSchnorrSig) error {
+	if ps.Index >= s.info.N {
+		return errors.New(fmt.Sprintf("Cannot add signature with index %d whereas schnorr could have max %s partial signatures", ps.Index, s.info.N))
+	}
+	if s.partials[ps.Index] != nil {
+		return errors.New(fmt.Sprintf("A Partial Signature has already been added for this index %d", ps.Index))
+	}
+	if err := s.verifyPartialSig(ps); err != nil {
+		return errors.New(fmt.Sprintf("Partial signature to add is not valid : %v", err))
+	}
+	s.partials[ps.Index] = ps
+	return nil
+}
+
+// SchnorrSig  will generate the global schnorr signature
+// By reconstructing the secret that the partial responses contains
+func (s *Schnorr) SchnorrSig() (*SchnorrSig, error) {
+	// automatic verification
+	// TODO : change this into a bool flag or public method ?
+	if err := s.verify(); err != nil {
+		return nil, err
+	}
+
+	pri := PriShares{}
+	pri.Empty(SUITE, s.info.T, s.info.N)
+	for i, ps := range s.partials {
+		pri.SetShare(ps.Index, *s.partials[i].Part)
+	}
+	// lagrange interpolation to compute the gamma
+	gamma := pri.Secret()
+	sig := &SchnorrSig{
+		Random:    s.random.Pub,
+		Signature: &gamma,
+	}
+	return sig, nil
+}
+
+// VerifySchnorrSig will verify if a given signature is correct regarding the message
+// NOTE: This belongs to the schnorr structs however it can be called at any time you want
+// This check is static, meaning it only needs the longterm shared secret, and the signature to
+// check. Think of the schnorr signature as a black box having two inputs
+//  - a message to be signed + a random secret ==> NewRound
+//  - a message + a signature to check on ==> VerifySchnorrSig
+func (s *Schnorr) VerifySchnorrSig(sig *SchnorrSig, msg []byte) error {
```

Comment:
Similarly, just 'Verify()'?  And take msg as a hash or something like that, not a byte-slice.